src/frontend/src/optimizer/plan_node/logical_source.rs

```diff
@@ -197,6 +241,73 @@ impl LogicalSource {
             kafka_timestamp_range: range,
         }
     }
+
+    #[must_use]
+    fn rewrite_to_stream_batch_source(&self) -> Self {
+        let column_catalog = self.core.column_catalog.clone();
+        // Filter out the generated columns.
+        let row_id_index = Self::rewrite_row_id_idx(&column_catalog, self.core.row_id_index);
+        let source_column_catalogs = column_catalog
+            .into_iter()
+            .filter(|c| !c.is_generated())
+            .collect_vec();
+        let core = generic::Source {
+            catalog: self.core.catalog.clone(),
+            column_catalog: source_column_catalogs,
+            pk_col_ids: self.core.pk_col_ids.clone(),
+            row_id_index,
+            gen_row_id: self.core.gen_row_id,
+            for_table: self.core.for_table,
+            ctx: self.core.ctx.clone(),
+        };
+        let base = PlanBase::new_logical_with_core(&core);
+
+        Self {
+            base,
+            core,
+            kafka_timestamp_range: self.kafka_timestamp_range,
+        }
+    }
+
+    fn wrap_with_optional_generated_columns_stream_proj(&self) -> Result<PlanRef> {
+        let column_catalogs = self.core.column_catalog.clone();
+        let pk_indices = self.logical_pk();
+        let exprs = Self::gen_optional_generated_column_project_exprs(
+            column_catalogs
+                .iter()
+                .map(|c| &c.column_desc)
+                .cloned()
+                .collect_vec(),
+        )?;
+        if let Some(exprs) = exprs {
+            let source = StreamSource::new(self.rewrite_to_stream_batch_source());
+            let logical_project = LogicalProject::with_pk_indices(source.into(), exprs, pk_indices);
+            Ok(StreamProject::new(logical_project).into())
```

Comment:
> Seems that the LogicalProject should use the pk that includes generated columns.

Yes

> if we do not update pk_col_ids in rewrite_to_stream_batch_source

We did not update pk_col_ids. And it is hard to avoid rewrite_to_stream_batch_source since since https://github.com/risingwavelabs/risingwave/pull/8841#discussion_r1152922155